exercises/rollback-with-saga/solution/src/workflows.ts

```
@@ -0,0 +1,92 @@
+import { proxyActivities, ApplicationFailure, ActivityFailure, log, sleep } from '@temporalio/workflow';
+import type * as activities from './activities';
+import { compensate, errorMessage } from './compensationUtils';
+import { Distance, PizzaOrder, OrderConfirmation, Compensation } from './shared';
+
+const { sendBill, getDistance, validateCreditCard, refundCustomer, updateInventory, revertInventory } =
+  proxyActivities<typeof activities>({
+    startToCloseTimeout: '5 seconds',
+    retry: {
+      maximumInterval: '10 seconds',
+    },
+  });
+
+export async function pizzaWorkflow(order: PizzaOrder): Promise<OrderConfirmation> {
+  let compensations: Compensation[] = [];
+  let totalPrice = 0;
+  
+  if (order.isDelivery) {
+    let distance: Distance | undefined = undefined;
+
+    try {
+      distance = await getDistance(order.address);
+    } catch (e) {
+      log.error('Unable to get distance', {});
+      throw e;
+    }
+    if (distance.kilometers > 25) {
+      throw ApplicationFailure.create({
+        message: 'Customer lives too far away for delivery',
+        details: [distance.kilometers],
+      });
+    }
+  }
+
+  for (const pizza of order.items) {
+    totalPrice += pizza.price;
+  }
+
+  // Validate the credit card number
+  try {
+    await validateCreditCard(order.customer.creditCardNumber);
+  } catch (err) {
```

Need to rethrow this err, otherwise workflow won't fail.